src/Lean/Message.lean

```diff
@@ -505,6 +528,56 @@ def indentD (msg : MessageData) : MessageData :=
 def indentExpr (e : Expr) : MessageData :=
   indentD e
 
+/--
+Returns the character length of the message when rendered.
+
+Note: this is a potentially expensive operation that is only relevant to message data that are
+actually rendered. Consider using this function in lazy message data to avoid unnecessary
+computation for messages that are not displayed.
+-/
+private def MessageData.formatLength (ctx : PPContext) (msg : MessageData) : BaseIO Nat := do
+  let { env, mctx, lctx, opts, ..} := ctx
+  let fmt ← msg.format (some { env, mctx, lctx, opts })
+  return fmt.pretty.length
+
+
+/--
+Renders an expression `e` inline in a message unless it will exceed `maxInlineLength` characters, in
+which case the expression is indented on a new line.
+
+Note that the output of this function is formatted with preceding and trailing space included. Thus,
+in `m₁ ++ inlineExpr e ++ m₂`, `m₁` should not end with a space or new line, nor should `m₂` begin
+with one.
+-/
+def inlineExpr (e : Expr) (maxInlineLength := 30) : MessageData :=
+  .ofLazy
+    (fun ctx? => do
+      let msg := MessageData.ofExpr e
+      let some ctx := ctx? | return Dynamic.mk (" " ++ msg ++ " ")
+      if (← msg.formatLength ctx) > maxInlineLength then
+        return Dynamic.mk (indentD msg ++ "\n")
+      else
+        return Dynamic.mk (" " ++ msg ++ " "))
+    (fun mctx => instantiateMVarsCore mctx e |>.1.hasSyntheticSorry)
```

Comment:
Why are you not using the `MessageData.lazy` wrapper that hides the `Dynamic` stuff? Is it because you want to fail more gracefully when the `ctx?` is empty?

Maybe `MessageData.lazy` should take an optional parameter of type `Unit → MessageData`

Reply:
> Is it because you want to fail more gracefully when the `ctx?` is empty?

That was the intent, yes; since `MessageData.ofExpr` has a fallback in the case of a missing context, I was trying to make this function consistent with that behavior. If this isn't a contingency worth handling, I can switch this to use `MessageData.lazy`; alternatively, I can add an optional parameter `onMissingContext := fun _ => pure (.ofFormat "(invalid MessageData.lazy, missing context)")` to that function and use it to replicate the current behavior in the PR.

Reply:
I’d add an optional paramter to `lazy`, so that we can contain the `Dynamic` to that wrapper

Reply:
I've added this and rewritten `inlineExpr` accordingly